src/Error.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+{-|
+Module      : Error
+Description : An instantiation of 'Validation' with our error type.
+
+This specializes 'Data.Validation.Validation' to keep its errors in a 'NonEmpty' list
+and keep track of a 'Pn' for every error it logs. There is also some infrastructure
+around modified chaining/branching behaviours.
+-}
+
+module Error (module Error) where
+
+import Data.Functor.Alt
+import Data.List.NonEmpty as NE
+import Data.Validation as Error
+import Data.Proxy
+import Data.Reflection
+
+import Syntax.Untyped (Pn)
+
+-- Reexport NonEmpty so that we can use `-XOverloadedLists` without thinking.
+import Data.List.NonEmpty as Error (NonEmpty)
+
+type Error e = Validation (NonEmpty (Pn,e))
+
+throw :: (Pn,e) -> Error e a
+throw msg = Failure [msg]
+
+infixr 1 <==<, >==>
+
+-- Like 'Control.Monad.(>=>)' but allows us to chain error-prone
+-- computations even without a @Monad@ instance.
+(>==>) :: (a -> Error e b) -> (b -> Error e c) -> a -> Error e c
+f >==> g = \x -> f x `bindValidation` g
+
+(<==<) :: (b -> Error e c) -> (a -> Error e b) -> a -> Error e c
+(<==<) = flip (>==>)
+
+-- | If there is no error at the supplied position, we accept this result and
+-- do not attempt to run any later branches, even if there were other errors.
+-- (The second argument looks intimidating but it simply demands that each
+-- @'Error' e a@ branch is wrapped in 'A' before being passed to '(<!>)'.)
+notAtPosn :: Pn -> (forall s. Reifies s (Alt_ (Error e)) => A s (Error e) a) -> Error e a
+notAtPosn p = withAlt $ \case
+  Failure err -> if any ((p ==) . fst) err then id else const $ Failure err
+  res         -> const res
+
+-- | Wraps any functor in a type that can be supplied a custom 'Alt' instance. 
+newtype A s f a = A { runA :: f a }
+  deriving Functor
+
+-- | The type of custom 'Alt' instances.
+newtype Alt_ f = Alt_ { alt :: forall a. f a -> f a -> f a }
+
+-- | Given a proof that we can reify a custom 'Alt' instance on the wrapped
+-- functor, we can give it an actual 'Alt' instance (allows using '(<!>)').
+instance (Functor f, Reifies s (Alt_ f)) => Alt (A s f) where
```

Comment:
[`Alt`](https://hackage.haskell.org/package/semigroupoids-5.3.6/docs/Data-Functor-Alt.html) is [`Alternative`](https://hackage.haskell.org/package/base-4.16.0.0/docs/Control-Applicative.html#t:Alternative) but without `empty`. I would've preferred to use `Alternative` since that's a more well-known class but `Validation` simply doesn't have such an instance (because you can't create one).

`Reifies` is kind of a hack. Maybe we should exclude it, since at this point it's only really used in `notAtPosn`, which could be implemented manually by a nested case analysis on our three known Act types. KISS, and all that.

I started explaining it but it turned waaay too long, which is definitely another argument against including it. But if you're curious, this is the tutorial I used: https://www.schoolofhaskell.com/user/thoughtpolice/using-reflection

Reply:
> Reifies is kind of a hack. Maybe we should exclude it, since at this point it's only really used in notAtPosn, which could be implemented manually by a nested case analysis on our three known Act types. KISS, and all that.

This sounds good to me.